### PR TITLE
Use provided target port in buildpack for 'containerapp up'

### PR DESCRIPTION
**Pull request by cormacpayne**
The current iteration of source-to-cloud using buildpacks in `containerapp up` is always listening to port 8080, independent of what the user is providing for the `--target-port` value.

This PR will propagate the `--target-port` value provided by the user to the buildpack, and ultimately the final run image, via the `PORT` environment variable that the buildpack is currently checking the value for (_i.e._, if `PORT` is set, used that as the port value in the final run image, otherwise use 8080).

EDIT: In addition to the above changes, this PR also introduces the following changes:
- In the `containerapp up` flow, when building a runnable application image using buildpacks, ensure that the user is authenticated against the created/provided ACR instance before pushing the produced image
- Add initial live-only tests for the `containerapp up` scenarios modified in this PR
  - _Note_: when an ACR Build/Task was executed by `containerapp up` (either when the user provided source with a Dockerfile or source without a Dockerfile that wasn't .NET 7), when using the Storage SDK to upload the source to a blob, there was a `utf8` error that would pop up with no way to resolve it. Looped in the Storage SDK for Python folks and their remediation was to update the Storage SDK used from V1 to V2, and doing so resolved the issue for one of the tests, however, the `upload_source_code` function that this logic is housed in was directly copied over from the `acr` module, and there is a flow in `containerapp up` that creates and runs an ACR Task which uses _their_ implementation of `upload_source_code` that still uses the old Storage SDK. I added the live-only test for the ACR Task scenario, but skipped it until we're able to update the `acr` module to use the new version of the SDK.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

`az containerapp up`

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

~- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).~

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

**Review by daniv-msft — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
Could you please add a test case for this PR change?

**Comment by cormacpayne** [comment]
@zhoxing-ms -- I just force pushed a commit with the previous target port change, along with some other changes I outlined in the "EDIT" of the PR description. These changes include live-only tests for this scenario, but I added context around why Storage SDK changes were needed within the `containerapp` module and future changes we'd need in the `acr` module that `containerapp` depends on. Please let me know if there are any additional questions.

**Comment by zhoxing-ms** [comment]
@cormacpayne Please add the PR description in the `HISTORY.rst`. Moreover, if you want release a new extension for this PR, please also upgrade the version defined in the `setup.py`.

**Comment by cormacpayne** [comment]
@zhoxing-ms -- I pushed a new commit to update the `HISTORY.rst` for `containerapp`, but before we merge this PR, I'm waiting to get a review and/or approval from additional Container Apps folks to ensure these changes are OK. I'll follow up when we should be good to merge this PR. Thanks!

File: src/containerapp/azext_containerapp/_archive_utils.py
Comment (daniv-msft) [inline src/containerapp/azext_containerapp/_archive_utils.py:56]: Could you please clarify where we're actually moving from V1 to V2? Here the code looks like what we need to update after we moved to the newer SDK, but I cannot find where the actual "Storage SDK update" happened. What am I missing?

File: src/containerapp/azext_containerapp/_up_utils.py
Comment (daniv-msft) [inline src/containerapp/azext_containerapp/_up_utils.py:508]: Just to validate: from your comment in the description of the PR it sound like the PORT env variable value is basically intercepted by our logic in the Buildpack and then used by us to set the port to expose on the final image. Is this correct?

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_up.py
Comment (daniv-msft) [inline src/containerapp/azext_containerapp/tests/latest/test_containerapp_up.py:46]: Please consider creating a helper function to share some of the logic between the test cases, as it looks like a good part is duplicated.

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_up.py
Comment (daniv-msft) [inline src/containerapp/azext_containerapp/tests/latest/test_containerapp_up.py:60]: Could you please clarify why we're not directly using a f-string rather than calling `format`?

File: src/containerapp/azext_containerapp/_archive_utils.py
Comment (Juliehzl) [reply src/containerapp/azext_containerapp/_archive_utils.py:56]: For resource type ResourceType.DATA_STORAGE_BLOB and DATA_STORAGE, they will correspond to different SDK.
https://github.com/Azure/azure-cli/blob/82358037c5af4a43c46facfcff054100a730c678/src/azure-cli-core/azure/cli/core/profiles/_shared.py#LL121-L126

Comment (cormacpayne) [reply src/containerapp/azext_containerapp/_archive_utils.py:56]: Yep, sorry, it's not obvious just from changing the `ResourceType` property, but the snippet that Zunli linked above is exactly what the Storage folks sent me when recommending to upgrade to their V2 SDK.

So we needed to change the `ResourceType` property so the `get_sdk` function targets the new SDK, and then we create a `BlobClient` object defined in the V2 Storage's `_blob_client.py` file.

File: src/containerapp/azext_containerapp/_up_utils.py
Comment (cormacpayne) [reply src/containerapp/azext_containerapp/_up_utils.py:508]: That's correct -- the `--env` argument will propagate the given environment variables to the stack that the detect/build phases occur, so our .NET 7 buildpack will have access to this `PORT` value set

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_up.py
Comment (cormacpayne) [reply src/containerapp/azext_containerapp/tests/latest/test_containerapp_up.py:60]: I was following the pattern in place for the original test in this suite, but I'll update the tests to use f-strings

**Review by daniv-msft — APPROVED** [review]
Looks good to me! Thanks

**Review by daniv-msft — APPROVED** [review]
Looks good to me! Thanks

**Comment by cormacpayne** [comment]
@zhoxing-ms -- We should be OK to merge this PR whenever someone from the CLI side has the opportunity to merge it.

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**
